clang/lib/DPCT/UserDefinedRules/UserDefinedRules.cpp

```diff
@@ -344,17 +344,138 @@ MetaRuleObject::PatternRewriter::PatternRewriter(
   Subrules = S;
 }
 
+// Read a YAML file recursively and substitute any "!include <filename>"
+// directive with the contents of the referenced file.
+std::unique_ptr<llvm::MemoryBuffer>
+readYAMLFile(const llvm::StringRef &RuleFilePath) {
+  // Load the rule file into a MemoryBuffer
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Buffer =
+      llvm::MemoryBuffer::getFile(RuleFilePath);
+  if (!Buffer) {
+    llvm::errs() << "Error: failed to read " << RuleFilePath << ": "
+                 << Buffer.getError().message() << "\n";
+    clang::dpct::ShowStatus(MigrationErrorInvalidRuleFilePath);
+    dpctExit(MigrationErrorInvalidRuleFilePath);
+  }
+
+  // Get the directory path of the rule file.
+  llvm::SmallString<128> DirectoryPath(RuleFilePath);
+  llvm::sys::path::remove_filename(DirectoryPath);
+
+  // Iterate over the input line by line.
+  std::stringstream Output, IncRuleFilePath;
+  size_t Ind = 0;
+  bool IncDirectiveFound = false;
+  bool SkipWhiteSpaces = false;
+  bool IsLineBeginning = true;
+
+  llvm::StringRef BuffContent = std::move(*Buffer)->getBuffer();
+  while (Ind < BuffContent.size()) {
+    char ch = BuffContent[Ind];
+
+    // Skip return carriage character
+    if (ch == '\r') {
+      Ind++;
+      continue;
+    }
+
+    // Skip white spaces at the beginning of the line if it contains a directive
+    if (IsLineBeginning) {
+      size_t i = Ind;
+      char c = ch;
+
+      // lookahead for "!" directive after white spaces
+      for (; i < BuffContent.size(); i++) {
+        c = BuffContent[i];
+        if (!(c == ' ' || c == '\t')) {
+          break;
+        }
+      }
+
+      // Move the Ind to the beginning of directive
+      if (c == '!' && i != Ind) {
+        Ind = i;
+        ch = c;
+      }
+
+      IsLineBeginning = false;
+    }
+
+    // Check if the line starts with a directive
+    if (ch == '!') {
+      // Check if the directive is "!include"
+      if (!IncDirectiveFound && BuffContent.substr(Ind, 8) == "!include") {
```

Comment:
need to check if the string has at least 7 chars after !

Reply:
Done.